File: backend/app/services/recurrence.py
from datetime import date
from typing import Any


def is_habit_due_on(rule: dict[str, Any] | None, d: date) -> bool:
    if not rule:
        return True
    t = rule.get("type", "daily")
    if t == "daily":
        return True
    if t == "weekly":
        days = rule.get("days", [])
        return d.isoweekday() in days
    return True

File: frontend/src/util/recurrence.js
/** @param {Record<string, unknown> | null | undefined} rule */
/** @param {Date} d local calendar date */
export function isHabitDueOn(rule, d) {
  const r = rule || {};
  const t = r.type || "daily";
  if (t === "daily") return true;
  if (t === "weekly") {
    const days = r.days || [];
    const jsDay = d.getDay();
    const iso = jsDay === 0 ? 7 : jsDay;
    return days.includes(iso);
  }
  return true;
}
